posthog/tasks/detect_alerts_anomalies.py

```diff
@@ -0,0 +1,77 @@
+from typing import cast
+
+import structlog
+from celery import shared_task
+from django.utils import timezone
+
+from posthog.email import EmailMessage
+from posthog.hogql_queries.legacy_compatibility.filter_to_query import filter_to_query
+from posthog.hogql_queries.query_runner import get_query_runner
+from posthog.models import Alert, AnomalyCondition
+from posthog.schema import HogQLQueryResponse
+
+logger = structlog.get_logger(__name__)
+
+
+def check_all_alerts() -> None:
+    alerts = Alert.objects.all().only("id")
+    for alert in alerts:
```

Comment:
I don't know for sure but this also feels like a scaling nightmare... We struggle sometimes to keep up with dashboard / insight refreshes in general and this is another form of refresh, just with a higher demand on reliability. I _think_ this would require strong co-ordination with @PostHog/team-product-analytics to make sure this fits in with their existing plans for improving background refreshing otherwise this will hit scaling issues fast.

Reply:
> I think this would require strong co-ordination with @PostHog/team-product-analytics to make sure this fits in with their existing plans for improving background refreshing otherwise this will hit scaling issues fast.

Just wanted to chime in here. I can take a look at this, but am currently busy with being on support for this sprint. I'll see what we can do.

Reply:
Scaling celery is not the issue, but ClickHouse will struggle and ultimately go down if suddenly 1000 simultaneous queries appear.

Reply:
> should scale horizontally pretty easily - add a separate queue for these events, increase the number of parallel tasks in flight and add more servers if needed.

yep, was going to add "Should" is doing a lot of work in this sentence 😅 

@webjunkie I'm too far removed from how query code and caching interacts here

we already have one set of jobs that (is|should be) staying on top of having insight results readily available. does this use that cache? we should really overlap them so we have one set of tasks keeping a cache warm and then another that reads the fast access data in that cache for anomaly detection

humans aren't visiting insights once a minute so we know this will generate sustained load.

we should totally, totally build this feature - it's long overdue

--- 

i'm not opposed to getting a simple version in just for our team or select beta testers so we can validate the flow, but this 100% needs an internal sponsor since the work of rolling this out and scaling it simply can't be given to an external contributor (it wouldn't be fair or possible 🙈)

i would _love_ to be the internal sponsor but it's both not possible and completely outside of my current wheelhouse

(these concerns might be addressed elsewhere - i've not dug in here at all 🙈)

Reply:
> but ClickHouse will struggle and ultimately go down if suddenly 1000 simultaneous queries appear

Can't I limit the number of celery queries in flight? I understand this will introduce a problem of throughput, but then if the servers can't process N alerts each hour, maybe more read replicas or more servers are needed. I don't have much experience with column oriented databases though, so it's just a speculation.

> we already have one set of jobs that (is|should be) staying on top of having insight results readily available. does this use that cache?

🤷 , well the query_runner has some "cache" substrings in it's code, so one could assume... But I don't know

> humans aren't visiting insights once a minute so we know this will generate sustained load.

Just to clarify, it's once an hour

> but this 100% needs an internal sponsor since the work of rolling this out and scaling it simply can't be given to an external contributor (it wouldn't be fair or possible 🙈)

I completely agree and I would be really happy to have a mentor on this task. 

BTW, an interesting data point - Mixpanel [limits](https://docs.mixpanel.com/docs/features/alerts) their number of alerts to 50 per project.

Reply:
We will talk among @PostHog/team-product-analytics next week and discuss this regarding ownership and so on.

Reply:
> > humans aren't visiting insights once a minute so we know this will generate sustained load.
> Just to clarify, it's once an hour

👍 

(same point but thanks for clarification :))